### PR TITLE
jax: batched point cloud processing

### LINKED ISSUE #29: `rsp.jax` special implementations (`CFAR`, `CFARCASO`, `PointCloud`) do not have a consistent API
Unlike the rest of `rsp.jax`, `CFAR`, `CFARCASO`, and `PointCloud` are not batched.
- We should either allow batches, or document that we expect the caller to `vmap`.

There's also a few ergonomic issues with the API and documentation:
- we should explicitly call out that `rsp.jax` supports additional features.
- many parameters like `range_resolution` in `PointCloud` should provide a default
- parameters should be better documented, including how they are used and their physical meanings
- axis names are not consistent (`ele` vs `el`, `azi` vs `az`)

### PR DESCRIPTION
**Pull request by huangjuite**
- Fix #29: unify all point cloud processing interfaces to process batched input.

Part of a stack (see #58): **#55 (this PR)** → #56 (torch backend) → #57 (tests).

### CODE REVIEW

**Comment by thetianshuhuang** [comment]
Can you TAL at the copilot comments? Many of them seem reasonable.

Some other notes:
- Would be good to get some unit tests as well.
- Since github has native stacked PR support now, can you try splitting this in two? I'd do the rsp.jax changes in PR 1, then rsp.torch in PR 2.
- We should also be able to get a numpy point cloud processing stack going as well for completeness.

File: src/xwr/rsp/jax/aoa.py [hunk 1]
```diff
@@ -32,98 +34,138 @@ class PointCloud:
...
 
         Returns:
-            ang: detect angle index for every range doppler bin.
+            ang: detect angle index for every range doppler bin, as
+                `(elevation, azimuth)` along the trailing axis.
         """
-        idxs = jax.vmap(jax.vmap(self._argmax_aoa))(cube)
+        idxs = jax.vmap(jax.vmap(jax.vmap(self._argmax_aoa)))(cube)
         ang = jnp.stack((idxs), axis=-1)
         return ang
```
[the remark this answers is not in the document]
Comment (thetianshuhuang) [reply src/xwr/rsp/jax/aoa.py]: Seems like a reasonable suggestion to me, actually cuts down on a lot of LOC and will save compile time at least

File: src/xwr/rsp/jax/aoa.py [hunk 2]
```diff
@@ -32,98 +33,142 @@ class PointCloud:
         scale factor when the chirp center frequency differs.
 
     Args:
-        range_resolution: range fft resolution
-        doppler_resolution: doppler fft resolution
-        angle_fov: angle field of view in degrees for (elevation, azimuth).
-        angle_size: angle fft size for (elevation, azimuth).
-        antenna_spacing: antenna spacing in terms of wavelength (default 0.5).
+        config: radar configuration; see [`XWRConfig`][xwr.config.]. Only the
```
Comment (thetianshuhuang) [inline src/xwr/rsp/jax/aoa.py:36]: Long writeups in the args section render weirdly and are generally hard to read since it makes it hard to get a high level view of all the args. Put the comprehensive writeup of the behavior in the main docstring, with the args section summarizing this.

File: src/xwr/rsp/jax/aoa.py [hunk 3]
```diff
@@ -32,98 +33,142 @@ class PointCloud:
...
+            estimated angle falls outside the bound are excluded from the
+            returned mask. This rejects estimates near the edge of the array's
+            sin-space, where a sparse MIMO array has little real resolving
+            power and grating lobes appear.
+        angle_size: angle fft size for (elevation, azimuth). **Must match the
+            `size={"elevation": ..., "azimuth": ...}` the
+            [`RSP`][xwr.rsp.] was constructed with**: the bin-to-angle lookup
+            tables are built at this length, then indexed by the argmax over
+            the cube's angle axes. A mismatch silently yields wrong angles
+            rather than an error.
```
Comment (thetianshuhuang) [inline src/xwr/rsp/jax/aoa.py:57]: Please check all documentation for AI-isms.

This is one egregious example, since if a mismatch silently yields wrong angles rather than an error, that sounds like a bug that we should fix!

This comment applies broadly to all the documentation in this PR and stack.

File: src/xwr/rsp/jax/aoa.py [hunk 4]
```diff
@@ -32,98 +33,142 @@ class PointCloud:
...
+            positive; a non-positive value raises `ValueError`, while a wrong
+            (but positive) value gives systematically wrong angles, not an
+            error.
     """
 
     def __init__(
         self,
-        range_resolution: float,
-        doppler_resolution: float,
+        config: XWRConfig,
```
Comment (thetianshuhuang) [inline src/xwr/rsp/jax/aoa.py:69]: Why are we changing to XWRConfig as the input? This is a substantial programming model break, not just an API break, since this creates an interface dependency between `xwr.rsp` and the high-level radar API.

File: src/xwr/rsp/jax/spectrum.py [hunk 5]
```diff
@@ -32,66 +32,122 @@ class CFAR:
...
+        self, signal: Float[Array, "range doppler"]
+    ) -> Float[Array, "range doppler"]:
+        """Get the ring-averaged noise floor for a range-doppler image."""
+        # Jax currently only supports 'fill', but this should be changed to
+        # 'wrap' if they ever decide to add support; the training cell count
+        # is normalized out to compensate at the edges.
+        valid = convolve2d(jnp.ones_like(signal), self.mask, mode="same")
+        return convolve2d(signal, self.mask, mode="same") / valid
+
+    def _cfar(
```
Comment (thetianshuhuang) [inline src/xwr/rsp/jax/spectrum.py]: Why is this helper even necessary, why can't we inline it?

File: src/xwr/rsp/jax/spectrum.py [hunk 6]
```diff
@@ -32,66 +32,122 @@ class CFAR:
...
+        """Run 2D cell-averaging CFAR.
 
         !!! note
 
-            Boundary cells are zero-padded.
+            The transmit and receive antenna axes are combined
+            non-coherently, so their relative order does not matter.
 
         Args:
-            x: input. If more than 2 axes are present, the additional axes
```
Comment (thetianshuhuang) [inline src/xwr/rsp/jax/spectrum.py:70]: Is there a reason why we're removing support for this?

**Review by thetianshuhuang — CHANGES_REQUESTED** [review]
This PR has a lot of issues. Please review carefully for AI-generated hallucinations and anti-patterns. There are probably more latent issues that I've missed in this first pass; I'll have to read through carefully again on revision.

File: src/xwr/rsp/jax/aoa.py (hunk 3 above)
Comment (huangjuite) [reply src/xwr/rsp/jax/aoa.py:57]: This one is fixed in the PR stack 4 #59

File: src/xwr/rsp/jax/aoa.py (hunk 4 above)
Comment (huangjuite) [reply src/xwr/rsp/jax/aoa.py:69]: you said you want parameters like range_resolution in PointCloud should provide a default, there is no such thing as a default. You need to provide it according to your modulation and FFT config. So I think providing config directly is more clear.

**Comment by thetianshuhuang** [comment]
Just skimmed through the rest of the stack; a suggested structural change:

- PR 1: Depending on what makes more sense, add interface(s) or base class(es) to `xwr.rsp` similar to the `xwr.rsp.RSP` base class. This will also help a lot with defining the right interface.
- PR 2: Make `rsp.jax` use the interface from PR 1
- PR 3: Add `rsp.numpy` and parity tests
- PR 4: Add `rsp.torch`

File: src/xwr/rsp/jax/spectrum.py (hunk 5 above)
Comment (huangjuite) [reply src/xwr/rsp/jax/spectrum.py]: oh it was vmapped directly, let me change it back.

File: src/xwr/rsp/jax/spectrum.py (hunk 6 above)
Comment (huangjuite) [reply src/xwr/rsp/jax/spectrum.py:70]: cause now the input is directly using post all FFT shape without the need to reshape to RD axis leading (r, d, ...)before passing to this function
https://github.com/RadarML/xwr/blob/2b81ed4a8c9461feecdaad44f4c71e0e5980a02f/demo/rerun_pc.py#L79-L87
